RPC Streaming Refactor

banks opened the pull request:
**Note this is built on the now very-out-of-date `streaming-rpc` branch**. A lot of work needs to happen to get this merged with Namespaces and other big changed happened in the last few months but I wanted to put this up for review on its own merit first.

This PR covers quite a lot of issues and a significant redesign of the streaming code. I'll describe the major (and minor) issues it solves first.

## Major Issues Solved

### 1. Missing Events

Previously the implementation could miss messages resulting in incorrect results. The issue was something we actually considered in design but then missed that some simplifications during implementation (i.e. not implementing buffering) re-introduced the bug. The test `TestStreaming_DeliversAllMessages` was written first to demonstrate the failure and failed consistently before this PR.

The issue stems from the following sequence of events:

 1. New Subscribe call for a topic start
 2. Snapshot of that topic is loaded from the current state store (index 10)
 3. While that snapshot is being encoded/delivered, a mutation changes the result in the state store (index 11) (e.g. new service instance registered)
 4. Subscribe endpoint finishes sending snapshot at index=10 and starts watching the publisher for any _new_ updates
 5. Client never sees the registration at index 11 and has an incomplete result indefinitely until the connections fails and it reconnects. It may even _still_ be inconsistent if other events have been delivered and may be until that client restarts.

The fix for this is not hard to see: we need to start watching for updated _before_ we generate the snapshot so that we can send any that come in during snapshot delivery. The problem here is that it implies buffering events, which in turn forces hard choices: how bug do we let the buffers grow? If they are limited then some size of result/speed of client combination is going to hit that limit and never be able to work. If we don't limit then a slow client can cause OOM on a server in extreme cases. This is made worse if you take the naive approach of just appending to a slice which auto-grows as then you can't free that memory even after you start delivering them until the client has caught up with the whole thing which might in bad cases be never.

It's also potentially wasteful if every subscribe call has its own buffer of the same set of events and it' own copy of those events.

The new buffer design solves this. More detail below.

### 2. Multiple events published under the same raft index could be missed by clients

Originally in the design we made a simplifying assumption that for now there would only ever be one event per topic per raft apply. In implementation we quickly hit the `/txn` endpoint which even in our limited roll out can already register a batch of service instances in one raft apply. We relaxed the one-event-per-index rule to accommodate.

But this causes another bug in the protocol. For example:

 1. A client subscribes and gets a snapshot of service instances at index 10
 2. Something (e.g. consul-k8s sync) registers 5 new instances of the service in one transaction at index 20
 3. The first registration event with `Index == 20` is sent to the client, but then the TCP connection breaks before the rest are received
 4. The client reconnects, thinking it's received everything up to and including index 20 so sets `SubscribeRequest.Index == 20`
 5. The server has no new events _after_ 20 and so assumes the client had all 5 of the events at 20

To work around this I added `EventBatch` message so a single event per-index-per-topic is still always true, but we can publish multiple events per topic per index and just batch them inside a single event wrapper so they are delivered atomically.

### 3. Applying `bexpr.Filter` on server side was incorrect

In the implementation so far we spent a bunch of cycles trying to figure out an efficient way for the server to filter events before sending them. At first glance this seems easy but then you have to situations like:

 1. Subscribe starts with `Filter == "Service.Tags contains foo"`
 2. Snapshot is sent that includes an instance `node1.service1` that has that tag
 3. That agent removes the tag from the instance in a subsequent `Register` call which we see in the topic stream
 4. The new register call doesn't match the filter so we ignore it...
 5. but now the client still has `node1.service1` stuck in its view of the result although it shouldn't be there any more.

We tried to mitigate this before by storing the set of unique event IDs we've sent and issuing removals if we filter an event with the same "unique key" later, but it made lots of assumptions about the nature of the events and the cardinality of data that would bite us later. I still found ways it could do the wrong thing too so I (speaking with Matt) decided it's simpler to just filter client side and put up with the penalty of sending stuff over the wire that might not be needed for now. It's likely cheaper for the server to do that in typical use-cases than maintain sufficient state for each subscriber that it can make intelligent decisions about what to send to maintain correct state at the other end.

## Minor issues addressed in the process

### 4. Distinguishing between an empty snapshot and no snapshot needed because you were up to date.

When a client reconnects to a server after a connection drop, it already has a snapshot/view of the state of the stream. Let's say it's already got a good copy of the state up to index 100. When it reconnects, it specifies `SubscribeRequest.Index = 100`. If the first message it gets back from the server is `EndOfSnapshot` there are two possible reasons for this:
 1. It's still up to date and there is nothing beyond index 100 to send so no new snapshot needed.
 2. It's totally out of date and the server _has_ sent a whole new snapshot, but it just happens to be an empty one (say because all the instances of that service deregistered).

In case 1 the client can just carry on and be correct, in case 2 it needs to reset its state entirely to be correct going forward. It was _technically_ possible to distinguish these cases before - in the first one the Index of the `EndOfSnapshot` event would match the request index and in the second it would be higher. But this felt very subtle (not to mention our own client code didn't handle this correctly.

While in here I added an explicit `ResumeStream` event to signal the first case so that if a client ever receives an `EndOfSnapshot` it knows it must replace it's current state with the events sent prior to that from the current call.

### 5. Added buffering

We left buffering as an "optimization" that helps when clients drop connections. The major issue 1 though pretty much required introducing it in some form though and the cleanest solution to all the other issues around that was essentially to build shared buffering so that's now done.

### 6. Add Snapshot Caching

Another "future" optimization we discussed was caching snapshots rather than regenerating them from the FSM. This is especially useful for thundering herds of new clients watching the same large result.

We punted on it because with most buffering designs it gets extremely complex managing the tradeoffs in memory usage vs CPU when deciding how long to keep snapshots cached for. Then you also have to tie that into the buffer size to make sure you still have all the events you need since that snapshot otherwise it becomes useless. Cache TTLs and buffer sizes all become tied to the "unknown" rate of updates for the topic etc.

As a byproduct of how I solved for problem 1 though, snapshot caching comes almost for free - it's almost no extra code and even better it can "auto tune" with very short cache times it will solve thundering herds nicely and no additional book-keeping overhead to see how long buffers need to be to keep it valid etc. I'm intersted to see if this has a big impact in benchmarks where we have an artificially large thundering herd.

## Not Fixed Here

There are a number of other issues I discovered while working on this that are filed in the project board but I've not tried to fix here to keep this PR somewhat (!) contained.

### Docs

Some documentation that needs updating - some bits of streaming were added to docs e.g. the `enable_backend_streaming` flag, but other parts are not mentioned at all or need additional updates. 

We might want a new page in architecture docs about it eventually although I don't think that's necessary for this release where it will be off-by default and somewhat experimental. 

I'd like to at least write up some developer docs for `contributing` both to explain how it works (and some of the content of this PR perhaps), but also create a checklist of how to port other endpoints to streaming so we can get more people working on that for next release.

### PreparedMessage

One optimisation that would be really easy now we have the buffer would be to protobuf encode the event once when we publish it to the buffer, and then deliver it to (possibly many) clients using protobuf's `PreparedMessage` which means you don't pay serialisation cost for every subscriber. Combined with snapshot caching this could really help with thundering herds of subscribers with large results and is not much more code but I want to get this in and benchmark it without first as a comparison.

### Bugs

I found 5 other bugs with the way health events are sent that could cause incorrect results. These mostly just need fixes in the way we send events rather than systemic changes to streaming so I've left them out. One is around ACL token expiry which is important but still can be done later.

## Buffer Design

The buffer design might not be super obvious to grok and I want to document it a bit so others are clear on how it works and why it's designed that way.

After considering many possible ways to implement this, I wanted to choose one that was both efficient and relatively simple to reason about. Performance was a secondary concern, but it did steer away from some naive approaches that would have pathological issue pretty quickly like leaking unbounded memory in worst case or synchronising all raft writes with all subscribers.

Goals were:
 - Wanted a single buffer per topic rather than every subscriber having their own buffer to minimise memory needed for thousands of subscribers
 - Don't want to have complex code/tuning to balance stability with large results or large churn with memory usage in the happy path. 
   - For example if you have fixed size buffers tuned to allow for "the worst possible" case of snapshot delivery time vs update rate, then either you set that low and cause instability in those freak moments when the limit is insufficient, or you tune high and waste a huge amount of memory buffering events no-one needs 99.999% of the time.
 - The events published during raft transaction should not block raft FSM apply
 - Thousands of subscribers must not arbitrarily delay message publishing by contending on a lock
 - Slow subscribers must not block publishing, but also may not have events dropped.

### Atomic Linked Lists

Consul's design relies heavily on a few properties of MemDB:

 1. It's an immutable radix tree which means that writes don't block reads since readers access the root element via an atomic pointer and then follow only internal pointers that are never modified.
 2. So readers have their "view" of the state store automatically kept alive for exactly the right amount of time for their read to complete, however slow they are, but as soon as it does, Go's garbage collector (GC) manages the cleanup without complex book keeping (in our code) around which tree nodes are being accessed still etc.
 3. So reads never need to take locks on state, and never block writes.

An original implementation of buffering I did was mostly lock-free but did use a lock to coordinate snapshotting and Armon raised this as a concern - even if it seemed quick most of the time it broken that nice property we have now that reads and snapshots never block writes or other readers.

We had then planned to solve this by building a buffer directly into memDB using an update I added last year that allowed range scans to find an index in an ordered queue. But that has it's own problems including potentially slowing down every Raft apply as we do book keeping on the queue, plus we have to decide on fixed size for the queue or do even more complex book keeping to figure out which events are still needed by current readers during a write which sounds horrible.

So instead I choose a data structure that has all the same properties as memdb listed above but without those downsides. It's a relatively simple singly-linked-list that using atomic pointers to allow concurrent append and tail without blocking. That sounds scary but it really is pretty simple.

The structure is called `stream.EventBuffer` and each list node just looks _kinda like_ this:

```
type BufferItem struct {
	Events []Event
	Err error
	next atomic.Value
	ch chan struct{}
}
```

Note that this is a slight simplification but it helps for now.

An `EventBuffer` itself is really just a pointer to one of these items which is the current "head" or most recent item. You can imagine it like this:

![image](https://user-images.githubusercontent.com/120915/73086781-2b0f6800-3ec9-11ea-8d07-e489e5ad2d8a.png)

 * Each arrow is an atomic pointer (actually an `atomic.Value` which is a not-unsafe way to do the same thing in Go).
 * Each item contains the set of events published during a single Raft apply to a single topic
 * The most recent event was published at raft index 16 and is being pointed to by the `EventBuffer` as the current head (most recent) event.
 * The `ch` chan in each item is closed and nil, except the item at the head. 
 * To append an item:
   *  Create new BufferItem will nil next pointer and a new chan (and containing the events obv.)
   * Update the `head` of the `EventBuffer` - all future readers looking for the "most recent" event will now see the new event
    * Update the `next` pointer of the previous head, all readers already iterating through from the old head node will now see the new item
    * close the chan on the old head, all readers who were waiting for the next event by blocking on that chan are now woken and follow the next pointer to see the new event
 * As soon as all readers have moved past an item in the buffer, Go can Garbage Collect (GC) it and free the memory so the buffer only remains as long as needed for the current readers just like memdb nodes.

### Snapshots

The flow of a streaming `Subscribe` call though doesn't only need to stream new events as they happen, but also needs to first deliver a "snapshot" of the current state for that query. For example in a service health query, the `Subscribe` streaming call must initially lookup all the currently registered instances for that service and send them to the client before it starts streaming the updates from that point.

The `EventBuffer` design actually has another really nice property - you can splice one `EventBuffer` on to the end of another since they are just linked lists which means the logic in the `Subscribe` endpoint can be really simple: start consuming a buffer that contains all the events in the snapshot, then when you get to the end, the latest events buffer can just be spliced straight on and the subscriber can continue following the same linked list to start streaming new events.

Having snapshots written into `EventBuffers` also means that they have the same auto-GC property - once the event has been sent to the subscriber it's no longer needed and will be GCed even before the whole snapshot has been sent.

This splicing trick enables the solution to problem 1:
 1. Initially, on a fresh `Subscribe` call, the current topic buffer's `head` can be stored. Any future updates made during snapshot delivery will now be captured by following the buffer from that item and the chain of pointers will keep them from being GCed until we are ready.
 2. a snapshot is generated from current FSM and streamed into a new (different) `EventBuffer`.
 3. When the whole snapshot has been written to the buffer, the topic buffer `head` item we stored in step 1 can now be "spliced" on to the end of the snapshot EventBuffer. This is conceptually the same as a regular `Append` with one slight subtlety I won't get into here.
 4. The `Subscribe` RPC handler (via `stream.Subscriber`) can then just follow the buffer sending events as fast as it sees them. It will keep sending right past the end of the snapshot and continue to iterate into the "new" events delivered during or after the snapshot was made.
    * note that it's possible some events after the `head` we stored were actually published before the snapshot and need to not be included again `Subscriber` can easily just skip them if their index is not greater than the snapshot index though!
  5. The subscriber eventually catches up with the latest event and then can sit and wait on the chan for the next update to the topic.
  
### Snapshot Caching

Now snapshots are just `EventBuffers` it's really easy to just cache the initial `head` item before the snapshot is written to in a map and have a new subscriber that needs the same snapshot start from there rather than stream a whole new snapshot from the state store.

Snapshot contents is kept around by the linked list pointers, but as soon as the original `head` is dropped from a cache (and all readers are done with it) GC will just clean up without any further accounting.

Caching is a tradeoff of space vs. cpu time generating snapshots. If you keep snapshots a long time, they not only take up memory themselves which is duplicating objects in the FSM, but they keep an increasingly long buffer of updates alive behind them.

Fortunately, creating snapshots is relatively cheap so the biggest benefit of caching at all is just de-duplicating work needed by multiple near-consecutive subscribers. For example if 1000 agents all start watching a service with 100 instances within a few seconds of each other, even a 10 second cache TTL reduces the number of FSM reads from 1000 to 1 and the number of FSM objects converted into events from 100,000 to 100.

So the TTLs are short and not-configurable for now as they are only meant to help de-duplicate work during thundering herds rather than provide long-term caches of bits of the FSM which it's not that expensive just to re-copy later if only one client needs it at a time.

s-christoff commented on agent/cache-types/materialized_view.go:
```diff
@@ -0,0 +1,419 @@
+type MaterializedView struct {
+	// Properties above the lock are immutable after the view is constructed in
+	// MaterializedViewFromFetch and must not be modified.
+	typ        StreamingCacheType
```
Is there a reason it is `typ` instead of `type`? 
It is _shorter_ but I feel like the addition of the extra letter could just help provide some clarity where I was like "what is a `typ`? Oh! "

banks replied:
Yeah it's gross but `type` is a reserved word in Go (see up there on line 73) so you can't use it for a field or variable name. We use `typ` everywhere else in the `cache` and `cache-types` package since Mitchell wrote it.

This is a bit like the ruby custom of naming things `klass` since `class` is a reserved word.

banks commented:
Found a bug with error handling: 
 - [ ] If the gRPC subscribe call errors, the view will log and backoff-loop, but it doesn't set any state which means that actual cache fetches will just timeout and get nil values back.

I think the best behaviour would be for the view to store the last error and close it's `updateCh` in this case so that any waiting `Fetch` calls will see the error and report that back to clients. In some cases that error might be transient (e.g. connection failure) and we could have waited it out and returned a good result. but if it's not (e.g. server rejecting request because of an encoding error or ACL violation) then we probably don't want to timeout.

Should try to mirror the behaviour of regular background-refresh blocking which I think would return the error from fetch and then let the cache return the last good value if there is one.

banks commented:
Fixed the error handling issue and added a test to cover it.

I also spent some time benchmarking locally the difference between snapshot caching and not since the difference in the overall benchmark was inconclusive.

I ran the simulated service watch workload against a local server on my laptop. I simulated 100 clients each with 100 watches (10k subscribers overall) watching one service with 2000 instances that was changing every second and initial ramp up of clients spread over 30 seconds. The server was able to process this without errors but using significant CPU. Measuring process CPU was not very reliable on a laptop doing many other things so was discounted.

These are the flame graphs of server process CPU usage during the 30 seconds where it is servicing new client snapshots with different snapshot caching options.

First with snapshot caching totally disabled (commented out in code so no map access penalties etc.:

![image](https://user-images.githubusercontent.com/120915/73285140-8012ed00-41ed-11ea-9fd6-b93180bda8c1.png)

 * 41% of the server's CPU samples were processing snapshots

With Snapshot caching and a TTL of 1 second:

![image](https://user-images.githubusercontent.com/120915/73285177-8f923600-41ed-11ea-8daf-2acecbd00f18.png)

 * doSnapshot doesn't even show up on the flame graph as it spent so little CPU time there
 * There are no other major CPU spikes here. Every major bit of CPU time is spend encoding results onto the wire or in Go's runtime.
 * worth noting that alone these flame graphs aren't necessarily conclusive as they don't show the total time taken to serve the workload (that was hard to measure accurately locally). What they do show though is that this optimization saves almost all of the CPU time in our code (not in gRPC handling or runtime) under this heavy thundering herd workload. Given other benchmarks that show overall performance is at least no worse this seems like a win.

With Snapshot caching and a TTL of 10 seconds:

![image](https://user-images.githubusercontent.com/120915/73285227-a042ac00-41ed-11ea-939d-84c167d272c1.png)


 * Pretty much indistinguishable from 1 second in this example where there is a huge thundering herd as thousands of snapshots are amortised over just a few calls to doSnapshot either way.
 * I think it's probably still better to use a slightly longer time like 10 seconds for more real-world workloads where there is likely to be more jitter over agents starting new blocking requests. The cost is minimal.

kyhavlov commented on agent/consul/stream/event_test.go:
```diff
@@ -0,0 +1,125 @@
+}
+func BenchmarkHealthSnapshotSnappy10000(b *testing.B) {
+	doBenchmarkHealthSnapshotSnappy(b, 10000)
+}
```
What did the comparisons look like here for snappy vs uncompressed?

kyhavlov commented on agent/consul/stream/subscription.go:
```diff
@@ -0,0 +1,114 @@
+
+		for _, e := range next.Events {
+			// Only return it if the key matches - the buffer is for the whole topic.
+			if s.req.Key == "" || s.req.Key == e.Key {
```
Will this have to change at all when we add topics like KV? Probably just adding a simple prefix check, right?

kyhavlov commented on agent/consul/streaming_endpoint.go:
```diff
@@ -28,210 +25,154 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
-		if _, ok := sent[idHash]; ok {
-			deleteEvent, err := stream.MakeDeleteEvent(&event)
-			if err != nil {
+		// TODO: bexpr filtering?
```
This is handled on the client side now, right?

banks commented:
Thanks for your review on this @kyhavlov.

I might actually abandon this PR in the sense that I'm now working on merging with master and the changes are so large it's actually ended up re-writing a good chunk of this to work slightly differently again but the comments here I'll go through and make sure they are taken care of in the merged branch.

This PR is still good one to review the changes to `EventBuffer` e.t.c. as none of that will change significantly, just the protobuf parts of `stream` and the state store integration points.

I'm also going to be simplifying a few things we are not actually using for now like the ACL rules being present in the responses etc.

banks commented on agent/cache-types/streaming_health_services.go:
```diff
@@ -36,7 +40,7 @@ func (c *StreamingHealthServices) Fetch(opts cache.FetchOptions, req cache.Reque
-	subscribeReq := stream.SubscribeRequest{
+	r := stream.SubscribeRequest{
 		Topic:      stream.Topic_ServiceHealth,
 		Key:        reqReal.ServiceName,
```
Yeah namespace support is coming in the merge branch. Actually first goal will be OSS merge with at least the right hook points for namespaces. I'll need another pass to actually make all of the streaming messages namespaceable later - we need to discuss how to do that as it involves potentially having different wire formats between OSS and Ent or including Namespace fields in protobufs in OSS.

banks commented on agent/consul/streaming_endpoint.go:
```diff
@@ -28,210 +25,154 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
-		if !allow {
-			allowEvent = false
+
+		filteredEvents, err := aclFilterEvents(events, aclFilter)
```
Interesting, I didn't know that was an option, but I don't think it "solves" the token expiration problem in the sense that we need to actively detect that and reset the stream so the client sees the permission error on a retry rather than just silently stops getting updates but will still serve the old values to the user that they aren't really allowed to any more via the agent cache.

I think token expiration is relatively easy though - we just need to create a `time.After` that forces a reset on the subscription at the expiry time.

banks commented:
Closed in favour of #7547
